### PR TITLE
Basic text wrapping support

### PR DESCRIPTION
**Pull request by socksy**
## Problem
- Text doesn't wrap very well, so it's hard to read the errors on a smaller screen
- Columns are fixed width, so difficult to zoom in on stuff
- The wrapping behaviour feels a bit buggy

## Solutions
- Let user choose per test report whether to wrap lines
- Add `resize: horizontal` to the column CSS on hover to enable basic resize (in the future this would feel a bit more professional if it were done on a large bar-like element, but perfect is the enemy of good)
- Fix misc CSS issues relating to wrapping

## Screenshots
<details><summary>Resizing</summary>

2021-02-25-003513_4xFeb_screenshot
</details>

<details><summary>Error messages don't overlap with the "ERROR" text</summary>

Before
2021-02-25-010240_4xFeb_screenshot
After
2021-02-25-003611_4xFeb_screenshot
</details>

<details><summary>The "ERROR"/"FAIL" text doesn't scroll with the error output anymore</summary>

Before
2021-02-25-010229_4xFeb_screenshot
After
2021-02-25-013400_4xFeb_screenshot

</details>

<details><summary>Each test report has a "wrap" checkbox that toggles line wrapping</summary>

2021-02-25-003531_4xFeb_screenshot
</details>

## Random Notes
The passing down of `wrap-lines?` everywhere feels non-ideal, and in general the state management feels a bit messy. Would you be amenable to a state management solution like re-frame (or another one)? If so I might spend the time at some point to adapt it.

Additionally, how would you feel about a refactor of the CSS? I think it could make sense to have less cascading, and possibly just inline styles in the components? It's quite hard to work out what parts will affect what right now. (although I suppose `deep-diff` might make it harder in that regard?)

### CODE REVIEW

File: modules/chui-ui/src/lambdaisland/chui/styles.clj [hunk 1]
```diff
@@ -111,7 +111,8 @@
                 :flex-direction :column
                 :padding ".5rem"
                 :overflow :auto}
-     [:&:hover {:background-color :snow}]]]
+     [:&:hover :&:active {:background-color :snow
```
Comment (socksy) [inline modules/chui-ui/src/lambdaisland/chui/styles.clj:114]: This I noticed when testing in chrome — sometimes when resizing a column it's no longer in `hover` state so flickers white and stops resizing. Not an issue in Firefox.

File: modules/chui-ui/src/lambdaisland/chui/ui.cljs [hunk 2]
```diff
@@ -341,19 +341,34 @@
           [:div :view-stacktrace
            (str "For stacktrace: See error number " error-number " in console")]))]])
 
+(defn assertions-header [title wrap-lines?-atom]
```
Comment (socksy) [inline modules/chui-ui/src/lambdaisland/chui/ui.cljs:344]: not sure if you have a convention here (I am quite partial to the pitch `!atom-name` convention, but know that it's possibly a bit contentious

Comment (plexus) [reply modules/chui-ui/src/lambdaisland/chui/ui.cljs:344]: no strong feelings either way

**Comment by plexus** [comment]
Thanks a lot @socksy! Great to see definite proof that someone is actually using this :D 

Released in `1.0.173`

I'm happy to consider improvements to the state management. I'm not generally as enthusiastic about re-frame as I once was though, and for a small project like this one I like the simplicity of just using Reagent. I think there's also a practical issue with re-frame, since in re-frame everything is a singleton we would potentially conflict with code under test that uses re-frame. It might also flat-out break things, e.g. IIRC when using re-posh the re-frame db actually becomes a Datascript db reference. Just some things to watch out for.

Regarding the CSS, I think if you see room for improvement we'd be very glad to take it. But I do have to defer to @barrosfelipe since he did the bulk of the frontend work here, maybe he has some specific considerations.

**Comment by socksy** [comment]
I'm no super fan of re-frame, but I think there's something to be said about having a clear separation of state concerns and view concerns. While the library of reagent is simpler, I find that having to constantly worry about where state is coming from, who knows about it, and where it gets set makes the app itself more complex.

Perhaps there's another, non-re-frame approach these days?

**Comment by plexus** [comment]
We really only have two state atoms, `runner/state` which is used by the test runner to collect test results and coordinate the test process, and `ui/ui-state`, which is essentialy `re-frame.db/app-db`. `show-passing?` and `wrap-lines?` are component-local state, those would still be there in re-frame.

**Comment by socksy** [comment]
If this were a re-frame app, I'd move those out from component level state and into the app-db, which would allow for easy refactorings (what if we want to change the wrapping state to be for the entire page, instead of a single test? what if "show-passing?" should be reflected in the individual test results? In re-frame the indirection between a subscription and the state means that you'd not have to change any of the view code). 

I'd likely keep the runner/state out of the app-db though, since only UI state would be there.

But as I said, I've no strong love for re-frame. Maybe worth mulling over some alternative solutions.

**Pull request merged**